libcudacxx/include/cuda/__warp_communication/warp_shuffle.h

```diff
@@ -0,0 +1,235 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPO__RATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___WARP_COMMUNICATION_SHFL_H
+#define _CUDA___WARP_COMMUNICATION_SHFL_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__cmath/ceil_div.h>
+#include <cuda/__ptx/instructions/get_sreg.h>
+#include <cuda/__ptx/instructions/shfl_sync.h>
+#include <cuda/std/__bit/has_single_bit.h>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__memory/addressof.h>
+#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/integral_constant.h>
+#include <cuda/std/__type_traits/is_pointer.h>
+#include <cuda/std/__type_traits/is_void.h>
+#include <cuda/std/cstdint>
+
+#if __cccl_ptx_isa >= 600
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <typename _Tp>
+struct WarpShuffleResult
+{
+  _Tp data;
+  bool pred;
+
+  template <typename _Up = _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE
+  operator cuda::std::enable_if_t<!cuda::std::is_array_v<_Up>, _Up>() const
+  {
+    return data;
+  }
+};
+
+template <int _Width = 32, typename _Tp>
+_CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE WarpShuffleResult<_Tp> warp_shuffle_idx(
+  const _Tp& __data, int __src_lane, uint32_t __lane_mask = 0xFFFFFFFF, _CUDA_VSTD::integral_constant<int, _Width> = {})
+{
+  constexpr int __warp_size   = 32;
+  constexpr int __is_void_ptr = _CUDA_VSTD::is_same_v<_Tp, void*> || _CUDA_VSTD::is_same_v<_Tp, const void*>;
+  static_assert(!_CUDA_VSTD::is_pointer_v<_Tp> || __is_void_ptr,
```

Comment:
To be sure, is that condition correct? or should it be
```suggestion
  static_assert(_CUDA_VSTD::is_pointer_v<_Tp> && __is_void_ptr,
```

Reply:
yes, it is the desidered behavior (see the documentation preview attached). My idea is to exchange any kind of data between threads, even pointers. 
The problem with pointers is that arrays can decay. I used this check to prevent user side errors. This is the same strategy used by `fmt::print`